ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs

```diff
@@ -796,8 +816,16 @@ rejectInvalidBlocks tracer registry getIsInvalidBlock getCandidate =
 -- the state explicit in the types and do the check in 'continueWithState'.
 newtype Stateful m blk tip s st = Stateful (s -> m (Consensus st blk tip m))
 
-continueWithState :: s -> Stateful m blk tip s st -> m (Consensus st blk tip m)
-continueWithState s (Stateful f) = f s
+continueWithState :: NoUnexpectedThunks s
+                  => s -> Stateful m blk tip s st -> m (Consensus st blk tip m)
+#if CHECK_TVAR_INVARIANT
```

Comment:
Thinking out loud: can't we have a single location where we have this `CPP` condition? Currently, this is the only location in `ouroboros-consensus` where we use it, but might that change?

Reply:
Let's come back to that once we have a second location :)

Reply:
That said, if you can think of any places where we have additional implicit state..?